### PR TITLE
[CELEBORN-1483] Add storage policy

### PR DESCRIPTION
**Pull request by FMX**
### What changes were proposed in this pull request?
To refactor partition data writer.
Part of CIP-8.

### How was this patch tested?
GA

### CODE REVIEW

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 1]
```diff
@@ -2814,6 +2827,30 @@ object CelebornConf extends Logging {
...
+      .checkValue(
+        _.split(",").map(str => StorageInfo.typeNames.contains(str.trim.toUpperCase)).forall(p =>
+          p),
+        "Will use default create file order. Default order: MEMORY,SSD,HDD,HDFS")
+      .createOptional
+
+  val WORKER_STORAGE_EVICT_POLICY: OptionalConfigEntry[String] =
+    buildConf("celeborn.worker.storage.storagePolicy.evictPolicy")
+      .categories("worker")
+      .doc("This define the order of evict files if the storages are available. Available storages: MEMORY,SSD,HDD,HDFS. Definition: StorageTypes|StorageTypes|StorageTypes.")
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2845]: eg: MEMORY,SSD|SSD,HDFS

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StoragePolicy.scala [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    this.createFileOrder = conf.workerStoragePolicyCreateFilePolicy
+    this.evictFileOrder = conf.workerStoragePolicyEvictFilePolicy
+    this.source = abstractSource
+  }
+
+  def getEvictedFile(
+      celebornFile: CelebornFile,
+      partitionDataWriterContext: PartitionDataWriterContext): CelebornFile = {
+    if (evictFileOrder.isDefined) {
+      if (evictFileOrder.get.contains(celebornFile.storageType.name())) {
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StoragePolicy.scala:48]: How about `val order = evictFileOrder.get.get(celebornFile.storageType.name())
if (null != order) {
    return createFile(partitionDataWriterContext, order)
}
null`

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StoragePolicy.scala [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      storageManager: StorageManager,
+      abstractSource: AbstractSource): Unit = {
+    this.storageManager = storageManager
+    this.conf = conf
+    this.createFileOrder = conf.workerStoragePolicyCreateFilePolicy
+    this.evictFileOrder = conf.workerStoragePolicyEvictFilePolicy
+    this.source = abstractSource
+  }
+
+  def getEvictedFile(
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StoragePolicy.scala:44]: Add UT for getEvictedFile

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StoragePolicy.scala [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+package org.apache.celeborn.service.deploy.worker.storage
+
+import org.apache.celeborn.common.CelebornConf
+import org.apache.celeborn.common.exception.CelebornIOException
+import org.apache.celeborn.common.internal.Logging
+import org.apache.celeborn.common.metrics.source.AbstractSource
+import org.apache.celeborn.common.protocol.StorageInfo
+
+object StoragePolicy extends Logging {
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StoragePolicy.scala:26]: IMO StoragePolicy belongs to StorageManager, better not expose StoragePolicy to CelebornFile.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/CelebornFileProxy.scala [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        currentFile.evict(nFile)
+        currentFile = nFile
+      }
+    }
+  }
+
+  def close(): Unit = {}
+
+  def isMemoryShuffleFile: Boolean = {
+    currentFile.isInstanceOf[CelebornMemoryFile]
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/CelebornFileProxy.scala:55]: Check storage type  instead using isInstanceOf

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StoragePolicy.scala (hunk 2 above)
Comment (FMX) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/storage/StoragePolicy.scala:48]: This change will throw NoSuchElementException.

**Comment by RexXiong** [comment]
Thanks, merge to main(v0.6.0)

**Pull request closed without merging**